src/network/serverpackethandler.cpp

```diff
@@ -1319,34 +1319,65 @@ void Server::handleCommand_Interact(NetworkPacket* pkt)
 
 			// Do stuff
 			pointed_object->rightClick(playersao);
-		}
-		else if (m_script->item_OnPlace(
-				item, playersao, pointed)) {
-			// Placement was handled in lua
 
-			// Apply returned ItemStack
-			if (playersao->setWieldedItem(item)) {
-				SendInventory(playersao);
-			}
+			return;
 		}
 
-		// If item has node placement prediction, always send the
-		// blocks to make sure the client knows what exactly happened
-		RemoteClient *client = getClient(pkt->getPeerId());
-		v3s16 blockpos = getNodeBlockPos(floatToInt(pointed_pos_above, BS));
-		v3s16 blockpos2 = getNodeBlockPos(floatToInt(pointed_pos_under, BS));
-		if (!item.getDefinition(m_itemdef).node_placement_prediction.empty()) {
-			client->SetBlockNotSent(blockpos);
-			if (blockpos2 != blockpos) {
-				client->SetBlockNotSent(blockpos2);
-			}
-		}
-		else {
-			client->ResendBlockIfOnWire(blockpos);
-			if (blockpos2 != blockpos) {
-				client->ResendBlockIfOnWire(blockpos2);
-			}
+		// Get the nodes before placing
+		bool under_ok, above_ok;
+		MapNode oldnode_u = m_env->getMap().getNodeNoEx(p_under, &under_ok);
+		MapNode oldnode_a = m_env->getMap().getNodeNoEx(p_above, &above_ok);
+		if (!under_ok || !above_ok) {
+			infostream << "Server: Not finishing placing: Node not found."
+				" Adding blocks to emerge queue." << std::endl;
+			m_emerge->enqueueBlockEmerge(peer_id,
+				getNodeBlockPos(p_under), false);
+			m_emerge->enqueueBlockEmerge(peer_id,
+				getNodeBlockPos(p_above), false);
+			return;
 		}
+
+		// The actual placing
+		if (!m_script->item_OnPlace(item, playersao, pointed))
+			// The on_place function was missing
+			return;
+
+		// Apply returned ItemStack
+		if (playersao->setWieldedItem(item))
+			SendInventory(playersao);
+
+		// Send current nodes to undo placement prediction if needed
+		if (item.getDefinition(m_itemdef).node_placement_prediction.empty())
+			return;
+
+		/*
+			There are several cases:
+			1. Client predicts under
+				i) Server predicts under
+					send under if the node didn't change
+				ii) Server predicts above
+					Client's under prediction must have already been undone
+			2. Client predicts above
+				i) Server predicts above
+					send above if the node didn't change
+				ii) Server predicts under
+					send above if the nodes didn't change
+
+			However, client prediction is unknown:
```

Comment:
?

Reply:
Why is the client prediction "unknown"?

Reply:
It's a synchronisation problem: The server only tests if the node at p_under is buildable to when the interact packet arrives, but the client may has put the node in an old version of the mapblock, or even put the node onto a wrongly predicted one. 
If the client puts two nodes, the second one onto the first one, and only the second one is placed, the server assumes that it's put in p_under (air is buildable_to, on_place actually puts the node there), whereas the client assumes that it's put on p_above (the firstly predicted node is not buildable_to).